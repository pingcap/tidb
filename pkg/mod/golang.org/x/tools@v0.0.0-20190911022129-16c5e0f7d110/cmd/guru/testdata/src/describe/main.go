package describe // @describe pkgdecl "describe"

// Tests of 'describe' query.
// See go.tools/guru/guru_test.go for explanation.
// See describe.golden for expected query results.

// TODO(adonovan): more coverage of the (extensive) logic.

import (
	"lib"
	"nosuchpkg"            // @describe badimport1 "nosuchpkg"
	nosuchpkg2 "nosuchpkg" // @describe badimport2 "nosuchpkg2"
	_ "unsafe"             // @describe unsafe "unsafe"
)

var _ nosuchpkg.T
var _ nosuchpkg2.T

type cake float64 // @describe type-ref-builtin "float64"

const c = iota // @describe const-ref-iota "iota"

const pi = 3.141     // @describe const-def-pi "pi"
const pie = cake(pi) // @describe const-def-pie "pie"
const _ = pi         // @describe const-ref-pi "pi"

var global = new(string) // NB: ssa.Global is indirect, i.e. **string

func main() { // @describe func-def-main "main"
	// func objects
	_ = main      // @describe func-ref-main "main"
	_ = (*C).f    // @describe func-ref-*C.f "..C..f"
	_ = D.f       // @describe func-ref-D.f "D.f"
	_ = I.f       // @describe func-ref-I.f "I.f"
	var d D       // @describe type-D "D"
	var i I       // @describe type-I "I"
	_ = d.f       // @describe func-ref-d.f "d.f"
	_ = i.f       // @describe func-ref-i.f "i.f"
	var slice []D // @describe slice-of-D "slice"

	var dptr *D // @describe ptr-with-nonptr-methods "dptr"
	_ = dptr

	// var objects
	anon := func() {
		_ = d // @describe ref-lexical-d "d"
	}
	_ = anon   // @describe ref-anon "anon"
	_ = global // @describe ref-global "global"

	// SSA affords some local flow sensitivity.
	var a, b int
	var x = &a // @describe var-def-x-1 "x"
	_ = x      // @describe var-ref-x-1 "x"
	x = &b     // @describe var-def-x-2 "x"
	_ = x      // @describe var-ref-x-2 "x"

	i = new(C) // @describe var-ref-i-C "i"
	if i != nil {
		i = D{} // @describe var-ref-i-D "i"
	}
	print(i) // @describe var-ref-i "\\bi\\b"

	// const objects
	const localpi = 3.141     // @describe const-local-pi "localpi"
	const localpie = cake(pi) // @describe const-local-pie "localpie"
	const _ = localpi         // @describe const-ref-localpi "localpi"

	// type objects
	type T int      // @describe type-def-T "T"
	var three T = 3 // @describe type-ref-T "T"
	_ = three

	print(1 + 2*3)        // @describe const-expr " 2.3"
	print(real(1+2i) - 3) // @describe const-expr2 "real.*3"

	m := map[string]*int{"a": &a}
	mapval, _ := m["a"] // @describe map-lookup,ok "m..a.."
	_ = mapval          // @describe mapval "mapval"
	_ = m               // @describe m "m"

	defer main() // @describe defer-stmt "defer"
	go main()    // @describe go-stmt "go"

	panic(3) // @describe builtin-ref-panic "panic"

	var a2 int // @describe var-decl-stmt "var a2 int"
	_ = a2
	var _ int // @describe var-decl-stmt2 "var _ int"
	var _ int // @describe var-def-blank "_"

	var _ lib.Outer // @describe lib-outer "Outer"

	var mmm map[C]D // @describe var-map-of-C-D "mmm"

	d := newD().ThirdField // @describe field-access "ThirdField"

	astCopy := ast
	unknown() // @describe call-unknown "\\("
}

type I interface { // @describe def-iface-I "I"
	f() // @describe def-imethod-I.f "f"
}

type C int
type D struct {
	Field        int
	AnotherField string
	ThirdField   C
}

func (c *C) f() {}
func (d D) f()  {}

func newD() D { return D{} }
