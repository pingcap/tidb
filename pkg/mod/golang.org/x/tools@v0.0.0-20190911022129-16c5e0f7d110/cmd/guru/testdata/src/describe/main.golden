-------- @describe pkgdecl --------
definition of package "describe"
	type  C      int
		method (*C) f()
	type  D      struct{...}
		method (D) f()
	type  I      interface{f()}
		method (I) f()
	const c      untyped int = 0
	type  cake   float64
	var   global *string
	func  main   func()
	func  newD   func() D
	const pi     untyped float = 3.141
	const pie    cake = 3.141

-------- @describe badimport1 --------
import of package "nosuchpkg"

-------- @describe badimport2 --------
reference to package "nosuchpkg"

-------- @describe unsafe --------
import of package "unsafe"
	builtin Alignof 
	builtin Offsetof
	type  Pointer  unsafe.Pointer
	builtin Sizeof  

-------- @describe type-ref-builtin --------
reference to built-in type float64

-------- @describe const-ref-iota --------
reference to const iota untyped int of value 0

-------- @describe const-def-pi --------
definition of const pi untyped float of value 3.141

-------- @describe const-def-pie --------
definition of const pie cake of value 3.141
Named types:
	type cake defined here

-------- @describe const-ref-pi --------
reference to const pi untyped float of value 3.141
defined here

-------- @describe func-def-main --------
definition of func main()

-------- @describe func-ref-main --------
reference to func main()
defined here

-------- @describe func-ref-*C.f --------
reference to method func (*C).f()
defined here

-------- @describe func-ref-D.f --------
reference to method func (D).f()
defined here

-------- @describe func-ref-I.f --------
reference to interface method func (I).f()
defined here

-------- @describe type-D --------
reference to type D (size 32, align 8)
defined as struct{Field int; AnotherField string; ThirdField C}
Methods:
	method (D) f()
Fields:
	Field        int
	AnotherField string
	ThirdField   C

-------- @describe type-I --------
reference to type I (size 16, align 8)
defined as interface{f()}
Methods:
	method (I) f()

-------- @describe func-ref-d.f --------
reference to method func (D).f()
defined here

-------- @describe func-ref-i.f --------
reference to interface method func (I).f()
defined here

-------- @describe slice-of-D --------
definition of var slice []D
Named types:
	type D defined here

-------- @describe ptr-with-nonptr-methods --------
definition of var dptr *D
Methods:
	method (*D) f()
Fields:
	Field        int
	AnotherField string
	ThirdField   C
Named types:
	type D defined here

-------- @describe ref-lexical-d --------
reference to var d D
defined here
Methods:
	method (D) f()
Fields:
	Field        int
	AnotherField string
	ThirdField   C
Named types:
	type D defined here

-------- @describe ref-anon --------
reference to var anon func()
defined here

-------- @describe ref-global --------
reference to var global *string
defined here

-------- @describe var-def-x-1 --------
definition of var x *int

-------- @describe var-ref-x-1 --------
reference to var x *int
defined here

-------- @describe var-def-x-2 --------
reference to var x *int
defined here

-------- @describe var-ref-x-2 --------
reference to var x *int
defined here

-------- @describe var-ref-i-C --------
reference to var i I
defined here
Methods:
	method (I) f()
Named types:
	type I defined here

-------- @describe var-ref-i-D --------
reference to var i I
defined here
Methods:
	method (I) f()
Named types:
	type I defined here

-------- @describe var-ref-i --------
reference to var i I
defined here
Methods:
	method (I) f()
Named types:
	type I defined here

-------- @describe const-local-pi --------
definition of const localpi untyped float of value 3.141

-------- @describe const-local-pie --------
definition of const localpie cake of value 3.141
Named types:
	type cake defined here

-------- @describe const-ref-localpi --------
reference to const localpi untyped float of value 3.141
defined here

-------- @describe type-def-T --------
definition of type T (size 8, align 8)
No methods.

-------- @describe type-ref-T --------
reference to type T (size 8, align 8)
defined as int
No methods.

-------- @describe const-expr --------
binary * operation of value 6

-------- @describe const-expr2 --------
binary - operation of value -2

-------- @describe map-lookup,ok --------
index expression of type (*int, bool)

-------- @describe mapval --------
reference to var mapval *int
defined here

-------- @describe m --------
reference to var m map[string]*int
defined here

-------- @describe defer-stmt --------
defer statement

-------- @describe go-stmt --------
go statement

-------- @describe builtin-ref-panic --------
function call (or conversion) of type ()

-------- @describe var-decl-stmt --------
definition of var a2 int

-------- @describe var-decl-stmt2 --------
definition of var _ int

-------- @describe var-def-blank --------
definition of var _ int

-------- @describe lib-outer --------
reference to type lib.Outer (size 56, align 8)
defined as struct{A int; b int; lib.inner}
No methods.
Fields:
	A                 int
	inner.C           bool
	inner.recursive.E bool

-------- @describe var-map-of-C-D --------
definition of var mmm map[C]D
Named types:
	type C defined here
	type D defined here

-------- @describe field-access --------
reference to field ThirdField C
defined here
Methods:
	method (*C) f()
Named types:
	type C defined here

-------- @describe call-unknown --------
function call of type invalid type

-------- @describe def-iface-I --------
definition of type I (size 16, align 8)
Methods:
	method (I) f()

-------- @describe def-imethod-I.f --------
definition of interface method func (I).f()

